examples/ctc.dx

```diff
@@ -13,89 +13,123 @@ However, Dex's autodiff can produce the backward
 pass automatically.  That makes this code much shorter than
 most implementations.
 
-import stats  -- for LogSpace
+import stats  -- for log-space representation of probabilities.
 
-'### Helper functions
+'## Custom index sets
 
-IsBlank = Bool
+'### Fence and Posts
+Here we create a custom datatype and index set
+that interleaves the elements of a table with another set
+of values representing all the spaces in between those elements.
 
-def interleave {m v} (blank:v) (labels: m=>v) : (m & IsBlank)=>v =
-  -- Turns "text" into "t e x t " by first pairing each letter with a blank,
-  -- then flattening the pairs back into a single-index table.
-  pairs = for i j. case j of
-    True  -> blank
-    False -> labels.i
-  for (i, j). pairs.i.j
+data FenceAndPosts n =
+  Fence n
+  Posts (Post n)
 
-def prepend {m v} (first: v) (seq: m=>v) : ((Unit|m)=>v) =
-  -- Concatenates a single element to the beginning of a sequence.
-  for idx. case idx of
-    Left () -> first
-    Right i -> seq.i
+instance {a} [Ix a] Ix (FenceAndPosts a)
+  size = 2 * size a + 1
+  ordinal = \i. case i of
+    Fence j -> 2 * ordinal j + 1
+    Posts j -> 2 * ordinal j
+  unsafe_from_ordinal = \o.
+    case is_odd o of
+      True  -> Fence $ unsafe_from_ordinal _ (intdiv2 o)
+      False -> Posts $ unsafe_from_ordinal _ (intdiv2 o)
+
+instance {a} [NonEmpty a] NonEmpty (FenceAndPosts a)
+  first_ix = unsafe_from_ordinal _ 0
+
+'### All but First
+Next we make an index set from any non-empty set
+that doesn't contain the first element.
+
+data AllButFirst n [NonEmpty n] =
+  MkAllButFirst n
+
+instance {a} Ix (AllButFirst a)
+  size = unsafe_nat_diff (size a) 1
+  ordinal = \(MkAllButFirst i). ordinal i
+  unsafe_from_ordinal = \o. MkAllButFirst $ unsafe_from_ordinal _ o
 
-def prepend_and_interleave {m v} (blank:v) (seq: m=>v) : ((Unit|(m & IsBlank))=>v) =
+instance {a} Subset (AllButFirst a) a
+  inject' = \x. unsafe_from_ordinal _ ((ordinal x) + 1)
+  project' = \x. case (ordinal x) < (size a) of
+    True  -> unsafe_from_ordinal _ (unsafe_nat_diff (ordinal x) 1)
+    False -> Nothing
```

Comment:
This doesn't look right.  You're checking that `ordinal x` is not too big, but then you're subtracting 1 from it?  Shouldn't this be `Nothing` if `ordinal x == 0`?

Let's write a little test of this index set and `Subset` instance (which you can either work into the demo or factor as a separate library with its own tests, or just embed here as an ugly test, possibly at the end).  `project . inject` should be `Just`, and `for x:(AllButFirst a). frob.(inject x)` should iterate over the desired set of elements of `frob : a => Int`.

Reply:
Great point, thanks for catching this.  I'll add a test to this example for now, and if we ever have another use for this index set, we can move it to the prelude or a library as you suggest.

Reply:
Fixed and tests added.